iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1043 @@
-                }
+                throw new InvalidOperationException($"Message size ({message.Payload.Length} bytes) is too big to process. Maximum allowed payload size is {MaxMessageSize}");
+            }
 
-                if (_completionQueue.Count == 0)
-                {
-                    throw new IotHubClientException("Unknown lock token.", isTransient: false);
-                }
+            var mqttMessage = new MqttApplicationMessageBuilder()
+                .WithTopic(TopicName)
+                .WithPayload(message.Payload)
+                .WithQualityOfServiceLevel(publishingQualityOfService)
+                .Build();
 
-                string actualLockToken = _completionQueue.Peek();
-                if (lockToken.IndexOf(actualLockToken, s_generationPrefixLength, StringComparison.Ordinal) != s_generationPrefixLength ||
-                    lockToken.Length != actualLockToken.Length + s_generationPrefixLength)
-                {
-                    throw new IotHubClientException(
-                        "Client must send PUBACK packets in the order in which the corresponding PUBLISH packets were received (QoS 1 messages) per [MQTT-4.6.0-2]. " +
-                        $"Expected lock token to end with: '{actualLockToken}'; actual lock token: '{lockToken}'.",
-                        isTransient: false);
-                }
+            MqttClientPublishResult result = await _mqttClient.PublishAsync(mqttMessage, cancellationToken);
 
-                _completionQueue.Dequeue();
-                completeOperationCompletion = _channel.WriteAndFlushAsync(actualLockToken);
+            if (result.ReasonCode != MqttClientPublishReasonCode.Success)
+            {
+                throw new IotHubClientException($"Failed to publish the mqtt packet for message with correlation id {message.CorrelationId} with reason code {result.ReasonCode}", true);
             }
-
-            await completeOperationCompletion.ConfigureAwait(true);
-
-            if (Logging.IsEnabled)
-                Logging.Exit(this, $"Completing a message with lockToken: {lockToken}", nameof(CompleteMessageAsync));
         }
 
-        public override Task AbandonMessageAsync(string lockToken, CancellationToken cancellationToken)
+        public override async Task SendEventAsync(IEnumerable<Message> messages, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            throw new NotSupportedException("MQTT protocol does not support this operation");
+            // Note that this sends all messages at once and then waits for all the acknowledgements. This
+            // is the recommended pattern for sending large numbers of messages over an asynchronous
+            // protocol like MQTT
+            List<Task> sendTasks = new List<Task>();
+            foreach (Message message in messages)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Why check on each message? Feels like that will just slow this operation down.

I'd check before, and then let Task.WhenAll below up (note it may throw an AggregateException or something and that will need to be managed.

Reply:
good idea